### DISCUSSION TITLE
Restricting symbolic constant values in grammar

### DISCUSSION THREAD

**Discussion by sushant94** [cvc5/cvc5#10201] [Q&A]
Hi,

Is there any way to restrict symbolic constant values in cvc5 grammar?

```python

bitvec = s.mkBitVectorSort(2)
boolean = s.getBooleanSort()


const_2 = s.mkVar(bitvec, 'const_2')
start = s.mkVar(boolean, 'start')

g = s.mkGrammar([], [start, const_2])

g.addAnyConstant(const_2)
g.addRules(start, [const_2])
```

In the above example, I want to add a constraint that const_2 cannot take a certain constant value, example, 0x0. Is there a way I can do that?

Alternatively, I'm interested in using this with SyGuS / generating abducts. So, any other way to emulate this behavior would also be helpful. Thanks! :)

**Comment by ajreynol** [comment]
At the moment, there isn't a good interface for this unfortunately.

The only feasible way I can see how to do this would be to use a deep embedding into a datatype, that is, synthesize a function over a grammar of datatype terms (`declare-datatype`) that are an embedding of arithmetic terms, where the semantics of the constructors is then given by a recursive evaluation function.  Then, one could specify custom constraints over the datatype to indicate that only the embedding of constants different than zero can be used. However, i expect this would incur significant performance overhead.

**Reply by sushant94** [reply]
Interesting. Do you have any pointers on how I could implement such an interface (within cvc5)? Or is this too big of a change?

It would be great if I could just express the above as a constraint on const_2 using something like an assertFormula. Let me know your thoughts.

**Comment by ajreynol** [comment]
The above suggestion regarding a deep embedding could already work out of the box, although the encoding would be tedious.

Another possibility, if you are interesting in only having *one* instance of `const_2` would be for it to be another function to synthesize.  Here is the idea:

```
(synth-fun c () Int)
(synth-fun f ((x Int)) Int
  ((Start Int))
  (
  (Start Int ((+ Start Start) c x 0 1))
  )
)
(check-synth)
```

Here, we are looking to synthesize `c` (which by virtue of having no arguments would require being a constant integer). Then, `f` would refer to `c`, with the restriction each copy of `c` generated by the grammar is the same value.

However, this is also disallowed currently in cvc5.  This is an extension to the sygus 2.0 language we call "forward declarations", see `fwd-decls` in https://saswat.padhi.me/assets/pdf/sygus_2.1.pdf.  This should be fairly easy to support.

However, this would also require further extensions to be allowed in the abduction feature, which assumes a single predicate to synthesize.

**Comment by sushant94** [comment]
Ok, that's encouraging. Yes, the const needs to be a fixed value for my use case, so forward declarations is exactly what I need. Is this something that's simple enough for me to implement?

I was a bit hesitant with the deep embedding solution because you mentioned significant performance overhead. The encoding shouldn't be a problem because I'm autogenerating these through python.

**Comment by ajreynol** [comment]
I opened a branch to add support for this: https://github.com/ajreynol/CVC4/tree/sygusFwdDecls, which wasn't too hard to do.

The branch requires a bit more fixes, but will hopefully be ready in the coming weeks.  I'll keep you posted.

**Reply by sushant94** [reply]
That's great! Thank you so much! This makes my life so much simpler :)

**Comment by ajreynol** [comment]
FYI, I've opened https://github.com/cvc5/cvc5/pull/10217.

**Reply by sushant94** [reply]
Awesome. Thanks a ton for the PR and merge!

Quick question: Does this already work with abducts? Or does something need to be changed?
